src/storage/entry.rs

```diff
@@ -0,0 +1,201 @@
+                storage: &mut *storage,
+            })
+        }
+    }
+
+    #[inline]
+    fn is_unconstrained() -> bool {
+        true
+    }
+}
+
+/// An entry to a storage which has a component associated to the entity.
+pub struct OccupiedEntry<'a, 'b: 'a, T: 'a, D: 'a> {
+    id: Index,
+    storage: &'a mut Storage<'b, T, D>,
+}
+
+impl<'a, 'b, T, D> OccupiedEntry<'a, 'b, T, D>
+where
+    T: Component,
+    D: Deref<Target = MaskedStorage<T>>,
+{
+    /// Get a reference to the component associated with the entity.
+    pub fn get(&self) -> &T {
+        unsafe { self.storage.data.inner.get(self.id) }
+    }
+}
+
+impl<'a, 'b, T, D> OccupiedEntry<'a, 'b, T, D>
+where
+    T: Component,
+    D: DerefMut<Target = MaskedStorage<T>>,
+{
+    /// Get a mutable reference to the component associated with the entity.
+    pub fn get_mut(&mut self) -> &mut T {
+        unsafe { self.storage.data.inner.get_mut(self.id) }
+    }
+
+    /// Converts the `OccupiedEntry` into a mutable reference bounded by
+    /// the storage's lifetime.
+    pub fn into_mut(self) -> &'a mut T {
+        unsafe { self.storage.data.inner.get_mut(self.id) }
+    }
+
+    /// Inserts a value into the storage and returns the old one.
+    pub fn insert(&mut self, mut component: T) -> T {
+        std::mem::swap(&mut component, self.get_mut());
+        component
+    }
+
+    /// Removes the component from the storage and returns it.
+    pub fn remove(self) -> T {
+        self.storage.data.remove(self.id).unwrap()
+    }
+}
+
+/// An entry to a storage which does not have a component associated to the entity.
+pub struct VacantEntry<'a, 'b: 'a, T: 'a, D: 'a> {
+    id: Index,
+    storage: &'a mut Storage<'b, T, D>,
+}
+
+impl<'a, 'b, T, D> VacantEntry<'a, 'b, T, D>
+where
+    T: Component,
+    D: DerefMut<Target = MaskedStorage<T>>,
+{
+    /// Inserts a value into the storage.
+    pub fn insert(self, component: T) -> &'a mut T {
+        self.storage.data.mask.add(self.id);
+        unsafe {
+            self.storage.data.inner.insert(self.id, component);
+            self.storage.data.inner.get_mut(self.id)
+        }
+    }
+}
+
+/// Entry to a storage for convenient filling of components or removal based on whether
+/// the entity has a component.
+pub enum StorageEntry<'a, 'b: 'a, T: 'a, D: 'a> {
+    /// Entry variant that is returned if the entity does has a component.
+    Occupied(OccupiedEntry<'a, 'b, T, D>),
+    /// Entry variant that is returned if the entity does not have a component.
+    Vacant(VacantEntry<'a, 'b, T, D>),
+}
+
+impl<'a, 'b, T, D> StorageEntry<'a, 'b, T, D>
+where
+    T: Component,
+    D: DerefMut<Target = MaskedStorage<T>>,
+{
+    /// Inserts a component if the entity does not contain a component.
+    pub fn or_insert(self, component: T) -> &'a mut T {
+        self.or_insert_with(|| component)
+    }
+
+    /// Inserts a component using a lazily called function that is only called
+    /// when inserting the component.
```

Comment:
I am not sure how to improve this, but it sounds wrong (to me)

Reply:
"Ensures this entry has a value and if not, inserts one using the result of the passed closure. Returns a reference to the value afterwards."

Reply:
Updated